File: sync_pbx.rb
require 'xcodeproj'
require 'pathname'

def sync_project(project_path, target_name)
  puts "Syncing #{project_path}..."
  project = Xcodeproj::Project.open(project_path)
  target = project.targets.find { |t| t.name == target_name }
  
  if target.nil?
    puts "Target #{target_name} not found!"
    return
  end

  source_dir = Pathname.new(project_path).dirname.join(target_name)
  
  # Find all physical Swift files (absolute paths)
  physical_files = Dir.glob("#{source_dir}/**/*.swift").map { |f| File.expand_path(f) }
  
  compile_phase = target.source_build_phase
  
  # 1. Remove missing or invalid files from compile phase FIRST
  compile_phase.files.to_a.each do |build_file|
    file_ref = build_file.file_ref
    if file_ref
      path = file_ref.real_path.to_s
      unless File.exist?(path)
        puts "Removing deleted/invalid file reference: #{path}"
        build_file.remove_from_project
        file_ref.remove_from_project
      end
    end
  end
  
  # 2. Get all valid files CURRENTLY in the compile phase
  existing_paths = []
  existing_basenames = []
  
  compile_phase.files.each do |file|
    path = file.file_ref&.real_path&.to_s
    if path
      existing_paths << File.expand_path(path).downcase
      existing_basenames << File.basename(path).downcase
    end
  end
  
  # 3. Add any physical files that are not in the compile phase
  physical_files.each do |absolute_path|
    basename = File.basename(absolute_path).downcase
    unless existing_paths.include?(absolute_path.downcase) || existing_basenames.include?(basename)
      # We need relative path to add to group
      file_path = Pathname.new(absolute_path).relative_path_from(Pathname.new(Dir.pwd)).to_s
      puts "Adding missing file: #{file_path}"
      
      # Add file to the project relative to the main group
      group_path = Pathname.new(file_path).dirname.relative_path_from(source_dir).to_s
      
      # Find or create group
      group = project.main_group.find_subpath(File.join(target_name, group_path), true)
      group.set_source_tree('<group>')
      
      file_ref = group.new_reference(file_path)
      compile_phase.add_file_reference(file_ref)
    end
  end
  
  project.save
  puts "Saved #{project_path}."
end

sync_project('Courier/Courier.xcodeproj', 'Courier')
sync_project('Customer/Customer.xcodeproj', 'Customer')

File: sync_pbx.py
import os
from pbxproj import XcodeProject

def sync_project(project_dir, project_name):
    pbx_path = os.path.join(project_dir, f"{project_name}.xcodeproj", "project.pbxproj")
    if not os.path.exists(pbx_path):
        print(f"Skipping {project_name}, not found.")
        return

    project = XcodeProject.load(pbx_path)
    
    # 1. Find all swift files currently in the project
    file_refs = project.objects.get_objects_in_section('PBXFileReference')
    
    # Track paths that are in the project to identify missing ones
    in_project_paths = []
    to_delete = []
    
    for obj in file_refs:
        # Check if it's a swift file
        path = getattr(obj, 'path', None)
        if getattr(obj, 'lastKnownFileType', None) == 'sourcecode.swift' or getattr(obj, 'explicitFileType', None) == 'sourcecode.swift' or (path and path.endswith('.swift')):
            pass
            
    # Actually, a better approach is to simply use the built-in methods.
    # Let's manually collect all physical .swift files
    physical_files = []
    for root, dirs, files in os.walk(os.path.join(project_dir, project_name)):
        for file in files:
            if file.endswith('.swift'):
                physical_files.append(os.path.join(root, file))
                
    # Now try to add each physical file. pbxproj `add_file` returns False if it's already there (usually).
    # We will pass force=False to avoid duplicates.
    print(f"Adding files for {project_name}...")
    for file_path in physical_files:
        # Add the file. We need to put it in the target.
        # But wait, pbxproj add_file doesn't create groups automatically if we don't specify the parent.
        # So we just add it to the project root group, or let XcodeProject do it.
        # Let's use add_file.
        # We can specify the tree as 'SOURCE_ROOT'
        # To avoid duplicating references, let's first check if the file name is in the project
        file_name = os.path.basename(file_path)
        existing = project.get_files_by_name(file_name)
        
        if not existing:
            print(f"Adding missing file: {file_name}")
            project.add_file(file_path, tree='SOURCE_ROOT', target_name=project_name)
            
    # Now for deletions
    print(f"Removing missing files for {project_name}...")
    for obj in project.objects.get_objects_in_section('PBXFileReference'):
        path = getattr(obj, 'path', None)
        if path and path.endswith('.swift'):
            # The path could be just the filename or a relative path.
            # Let's search if this filename exists in our physical files
            file_name = os.path.basename(path)
            found = any(f.endswith(file_name) for f in physical_files)
            if not found:
                print(f"Removing deleted file from project: {path}")
                project.remove_file_by_id(obj.get_id())
                
    project.save()
    print(f"Saved {project_name}.")

# Sync both projects
sync_project("Courier", "Courier")
sync_project("Customer", "Customer")
